packages/components/src/components/domainproperties/dataset/DatasetDesignerPanels.tsx

```diff
@@ -135,39 +144,59 @@ export class DatasetDesignerPanelImpl extends React.PureComponent<Props & Inject
 
-        // TODO: this rowIndexChange handles if the selected keyPropertyIndex or visitDatePropertyIndex field is moved but not if another field moves up above it
-
         this.setState(
             produce((draft: Draft<State>) => {
                 draft.model.domain = domain;
 
                 if (keyPropertyIndex !== undefined) {
                     // if the row was removed or reordered, update the keyPropertyIndex
-                    if (rowIndexChange && rowIndexChange.originalIndex === keyPropertyIndex) {
-                        draft.keyPropertyIndex = rowIndexChange.newIndex;
+                    if (rowIndexChange) {
+                        if (rowIndexChange.newIndex === 0 || rowIndexChange.newIndex) {
+                            domain.fields.map((field, index) => {
+                                if (this._selectedKeyFieldName && field.name === this._selectedKeyFieldName) {
+                                    draft.keyPropertyIndex = index;
+                                }
+                            });
+                            draft.model.keyPropertyName = domain.fields.get(draft.keyPropertyIndex).name;
+                        }
+                        // if row was removed, reset key property name
+                        else if (
+                            rowIndexChange.newIndex === undefined &&
+                            rowIndexChange.originalIndex === keyPropertyIndex
+                        ) {
+                            draft.model.keyPropertyName = '';
+                            draft.keyPropertyIndex = undefined;
+                        }
                     }
-
-                    // if row was removed, reset key property name
-                    if (draft.keyPropertyIndex === undefined) {
-                        draft.model.keyPropertyName = '';
-                    } else {
-                        // pick up any name changes to the selected field
-                        draft.model.keyPropertyName = domain.fields.get(draft.keyPropertyIndex).name;
+                    // pick up any name changes to the selected key field
+                    if (draft.keyPropertyIndex !== undefined) {
+                        this._selectedKeyFieldName = domain.fields.get(draft.keyPropertyIndex).name;
                     }
                 }
 
                 if (visitDatePropertyIndex !== undefined) {
                     // if the row was removed or reordered, update the visitDatePropertyIndex
-                    if (rowIndexChange && rowIndexChange.originalIndex === visitDatePropertyIndex) {
-                        draft.visitDatePropertyIndex = rowIndexChange.newIndex;
+                    if (rowIndexChange) {
+                        if (rowIndexChange.newIndex === 0 || rowIndexChange.newIndex) {
+                            domain.fields.map((field, index) => {
+                                if (this._selectedVisitDateName && field.name === this._selectedVisitDateName) {
```

Comment:
may be it can be solved without the private var but I used it because when we update the text of the field, I don't think model. visitDatePropertyName is updated at this point.

Reply:
I tried it without private var several times without changing much design and wasn't successful. Let me know if you've other ideas.